File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/VaultsGenerator.scala
package coop.rchain.casper.genesis.contracts

import coop.rchain.models.NormalizerEnv
import coop.rchain.rholang.build.CompiledRholangSource

final class VaultsGenerator private (supply: Long, code: String)
    extends CompiledRholangSource(code, NormalizerEnv.Empty) {
  val path: String = "<synthetic in VaultsGenerator.scala>"
}

object VaultsGenerator {

  // System vault initialization in genesis is done in batches.
  // In the last batch `initContinue` channel will not receive
  // anything so further access to `SystemVault(@"init", _)` is impossible.

  def apply(userVaults: Seq[Vault], supply: Long, isLastBatch: Boolean): VaultsGenerator = {
    val vaultBalanceList =
      userVaults.map(v => s"""("${v.vaultAddress.toBase58}", ${v.initialBalance})""").mkString(", ")

    val code: String =
      s""" new rl(`rho:registry:lookup`), systemVaultCh in {
         #   rl!(`rho:vault:system`, *systemVaultCh) |
         #   for (@(_, SystemVault) <- systemVaultCh) {
         #     new systemVaultInitCh in {
         #       @SystemVault!("init", *systemVaultInitCh) |
         #       for (TreeHashMap, @vaultMap, initVault, initContinue <- systemVaultInitCh) {
         #         match [$vaultBalanceList] {
         #           vaults => {
         #             new iter in {
         #               contract iter(@[(addr, initialBalance) ... tail]) = {
         #                  iter!(tail) |
         #                  new vault, setDoneCh in {
         #                    initVault!(*vault, addr, initialBalance) |
         #                    TreeHashMap!("set", vaultMap, addr, *vault, *setDoneCh) |
         #                    for (_ <- setDoneCh) { Nil }
         #                  }
         #               } |
         #               iter!(vaults) ${if (!isLastBatch) "| initContinue!()" else ""}
         #             }
         #           }
         #         }
         #       }
         #     }
         #   }
         # }
     """.stripMargin('#')

    new VaultsGenerator(supply, code)
  }
}

File: casper/src/rust/genesis/contracts/vaults_generator.rs
// See casper/src/main/scala/coop/rchain/casper/genesis/contracts/VaultsGenerator.scala

use super::vault::Vault;

pub struct VaultsGenerator {
    pub supply: i64,
    pub code: String,
}

impl VaultsGenerator {
    pub fn new(supply: i64, code: String) -> Self {
        Self { supply, code }
    }

    pub fn create_from_user_vaults(
        user_vaults: Vec<Vault>,
        supply: i64,
        is_last_batch: bool,
    ) -> Self {
        let vault_balance_list = user_vaults
            .iter()
            .map(|v| {
                format!(
                    "(\"{}\", {})",
                    v.vault_address.to_base58(),
                    v.initial_balance
                )
            })
            .collect::<Vec<String>>()
            .join(", ");

        let continue_clause = if !is_last_batch {
            "| initContinue!()"
        } else {
            ""
        };

        let code = format!(
            r#" 
            new rl(`rho:registry:lookup`), systemVaultCh in {{
              rl!(`rho:vault:system`, *systemVaultCh) |
              for (@(_, SystemVault) <- systemVaultCh) {{
                new systemVaultInitCh in {{
                  @SystemVault!("init", *systemVaultInitCh) |
                  for (TreeHashMap, @vaultMap, initVault, initContinue <- systemVaultInitCh) {{
                    match [{}] {{
                      vaults => {{
                        new iter in {{
                          contract iter(@[(addr, initialBalance) ... tail]) = {{
                          iter!(tail) |
                          new vault, setDoneCh in {{
                            initVault!(*vault, addr, initialBalance) |
                            TreeHashMap!("set", vaultMap, addr, *vault, *setDoneCh) |
                            for (_ <- setDoneCh) {{ Nil }}
                          }}
                      }} |
                      iter!(vaults) {}
                    }}
                  }}
                }}
              }}
            }}
          }}
        }}
      "#,
            vault_balance_list, continue_clause
        );

        Self::new(supply, code)
    }
}
